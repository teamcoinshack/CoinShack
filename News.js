import React, {Component} from 'react';
import {
  Text, 
  View, 
  ActivityIndicator, 
  StyleSheet, 
  FlatList, 
  Button,
  RefreshControl,
  TouchableOpacity,
  Image,
} from 'react-native';
import Firebase from 'firebase';
import q from './Query.js';

const background = '#373b48';
export default class News extends Component {
  constructor(props) {
    super(props);

    this.state = {
      id: '',
      news: [], 
      refreshing: false,
    }

    this.load = this.load.bind(this);
    this.refresh = this.refresh.bind(this);
    this.renderRow = this.renderRow.bind(this);
  }
  
  load(url) {
    this.props.navigation.navigate('InternalWebpage', {url: url});  
  }

  async refresh() {
    try {
      let articles = await q.getNews();
      articles = articles.filter(article =>
        (article.title.includes("oin") || article.title.includes("rypto"))
      )
      names = [];
      newArticles = [];
      articles.forEach(function(x) {
        if (!names.includes(x.title)) {
          names.push(x.title);
          newArticles.push(x);
        }
      });
      articles = newArticles;
      let count = 0;
      articles.map(function(x) {
        x.listId = count.toString();
        count++;
        return x;
      })
      this.setState({
        news: articles,
        refreshing: false,
      });
    } catch(error) {
      console.log(error);
    }
  }

  async componentDidMount() {
    try {
      this.setState({
        id: Firebase.auth().currentUser.uid,
        refreshing: true,
      });
      await this.refresh();
    } catch(error) {
      console.log(error);
    }
  }

  renderRow({item}) {
    return (
      <TouchableOpacity 
        style={styles.row}
        onPress={() => this.load(item.url)}
      >
        <View style={{ flexDirection: 'row' , justifyContent: 'center',}}>
          <View style={styles.imageContainer}>
            <Image
              source={{ uri: item.urlToImage }}
              style={styles.imageStyle}
            />
          </View>
          <View style={styles.textStyle}>
            <Text style={{ 
              fontSize: 18, 
              color: '#a4a9b9', 
              fontWeight: '500' 
            }}>
              {item.title}
            </Text>
          </View>
        </View>
      </TouchableOpacity>
    )
  }

  render() {
    if (this.state.refreshing) {
      return (
        <View style={styles.loadingStyle}>
          <View>
            <ActivityIndicator color="#999999" />
          </View>
        </View>
      )
    }
    return (
      <View style={styles.container}>
        <FlatList
          style={styles.flatStyle}
          data={this.state.news}
          renderItem={this.renderRow}
          keyExtractor={item => item.listId}
          refreshControl={
            <RefreshControl
              refreshing={this.state.refreshing}
              onRefresh={this.onRefresh}
            />
          }
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  loadingStyle: {
    flex: 1,
    backgroundColor: background,
    flexDirection: 'column',
    justifyContent: 'center',
  },
  row: {
    elevation: 1,
    borderRadius: 5,
    backgroundColor: '#515360',
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    paddingTop: 15,
    paddingBottom: 15,
    paddingLeft: 18,
    paddingRight: 16,
    marginLeft: 14,
    marginRight: 14,
    marginTop: 0,
    marginBottom: 12,
  },
  imageContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  imageStyle: {
    width: 80,
    height: 80,
  },
  textStyle: {
    flex: 1,
    paddingLeft: 18,
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: background,
  },
  cashText: {
    fontSize: 30,
    fontWeight: 'bold'
  },
  flatStyle: {
    marginTop: 20,
  },
});
