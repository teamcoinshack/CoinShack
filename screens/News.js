import React, { Component } from 'react';
import {
  Text, 
  View,
  ScrollView,
  StyleSheet, 
  FlatList, 
  RefreshControl,
  TouchableOpacity,
  Image,
  Dimensions,
} from 'react-native';
import Firebase from 'firebase';
import q from '../Query.js';
import MyBar from '../components/MyBar.js';
import RBSheet from 'react-native-raw-bottom-sheet';
import { coinTitles } from '../Masterlist.js';
import { CheckBox } from 'react-native-elements';

const background = '#373b48';

export default class News extends Component {
  constructor(props) {
    super(props);

    let coinFilters = coinTitles.reduce((acc, curr) => {
      return {...acc, [curr]: true};
    }, {});

    this.state = {
      id: '',
      news: [], 
      refreshing: false,
      allCrypto: true,
      ...coinFilters,
    }

    this.load = this.load.bind(this);
    this.refresh = this.refresh.bind(this);
    this.renderRow = this.renderRow.bind(this);
    this.onRefresh = this.onRefresh.bind(this);
    this.timePublished = this.timePublished.bind(this);
    this.toggleFilter = this.toggleFilter.bind(this);
    this.renderFilterSheet = this.renderFilterSheet.bind(this);
    this.getCheckBoxOnPress = this.getCheckBoxOnPress.bind(this);
    this.toggleAllCrypto = this.toggleAllCrypto.bind(this);
  }
  
  load(url) {
    this.props.navigation.navigate('InternalWebpage', {url: url});  
  }

  onRefresh() {
    this.setState({
      refreshing: true,
    }, function() { this.refresh() })
  }

  async refresh() {
    try {
      let topics = "";
      if (!(this.state.allCrypto)) { // get news of these cryptos only
        for (title of coinTitles) {
          if (this.state[title]) {
            topics += ("+" + title); // include this coin
          }
        }
      } else { // get news of all cryptos
        topics = "cryptocurrency";
      }

      let articles = await q.getNews(topics);

      titles = [];
      articles = articles.filter(x => {
        if (titles.includes(x.title)) {
          return false;
        } else {
          titles.push(x.title);
          return true;
        }
      });
      articles.forEach((x, index) => x.listId = "" + index);

      this.setState({
        news: articles,
        refreshing: false,
      });
    } catch(error) {
      console.log(error);
    }
  }

  async componentDidMount() {
    this.props.navigation.setParams({ toggleFilter: this.toggleFilter });

    try {
      this.setState({
        id: Firebase.auth().currentUser.uid,
        refreshing: true,
      });
      await this.refresh();
    } catch (error) {
      console.log(error);
    }
  }

  timePublished(date) {
    const published = new Date(date);
    const now = new Date();
    const milis = now.getTime();
    const hour = Math.round((milis - published) / 3600000);
    const day = Math.round((milis - published) / 86400000);
    if (hour === 0) {
      return 'now';
    } else if (hour === 1) {
      return '1 hour ago';
    } else if (hour < 24) {
      return hour + ' hours ago';
    } else if (day === 1) {
      return '1 day ago';
    } else {
      return day + ' days ago';
    }
  }

  // when allCrypto is false and toggled to true, set all coins to true
  // when allCrypto is true, only toggle itself to false
  toggleAllCrypto() {
    let newCoinStates = coinTitles.reduce((acc, curr) => {
      return {...acc, [curr]: !(this.state.allCrypto)};
    }, {});

    this.setState({
      allCrypto: !(this.state.allCrypto),
      ...newCoinStates,
    });
  }

  getCheckBoxOnPress(coinTitle) {
    return () => this.setState({ [coinTitle]: !this.state[coinTitle] });
  }

  toggleFilter() {
    this.RBSheet.open();
  }

  renderFilterSheet() {
    return (
      <RBSheet
        ref={ref => this.RBSheet = ref}
        height={Math.round(Dimensions.get('window').height) / 2 - 100}
        duration={250}
        onClose={this.refresh}
        customStyles={{
          container: {
            backgroundColor: background,
            alignItems: 'stretch',
          }
        }}
      >
        <ScrollView contentContainerStyle={styles.filterContainer}>
          <Text style={styles.filterSheetHeader}>
            Show news about
          </Text>
          <CheckBox
            title="All Cryptocurrencies"
            checked={this.state.allCrypto}
            onPress={this.toggleAllCrypto}
            containerStyle={{
              backgroundColor: "#515360",
              borderColor: "#515360",
            }}
            textStyle={{
              color: "#ffffff",
            }}
            checkedColor="#00f9ff"
          />
          {coinTitles.map((title, key) => {
             return (
              <CheckBox
                key={key}
                title={title}
                checked={this.state[title]}
                onPress={this.getCheckBoxOnPress(title)}
                containerStyle={{
                  backgroundColor: "#515360",
                  borderColor: "#515360",
                }}
                textStyle={{
                  color: "#ffffff",
                }}
                checkedColor="#00f9ff"
              />
            );
          })}
        </ScrollView>
      </RBSheet>
    );
  }

  renderRow({ item }) {
    return (
      <TouchableOpacity 
        style={styles.row}
        onPress={() => this.load(item.url)}
      >
        <View style={{ 
          flexDirection: 'row', 
          justifyContent: 'center',
        }}>
          <View style={styles.imageContainer}>
            { item.urlToImage === null || item.urlToImage === ''
              ? (
                <Image
                  source={require('../assets/icons/news.png')}
                  style={styles.imageStyle}
                />
              )
              : (
                <Image
                  source={{ uri: item.urlToImage}}
                  style={styles.imageStyle}
                />
              )
            }
          </View>
          <View style={styles.textStyle}>
            <Text style={{ 
              fontSize: 18, 
              color: '#bec4d8', 
              fontWeight: '500' 
            }}>
              {item.title}
            </Text>
          </View>
        </View>
        <View style={{
          flexDirection: 'row',
          justifyContent: 'space-between',
          paddingTop: 5,
        }}>
          <View style={styles.sourceStyle}>
            <Text style={{
              fontSize: 13,
              color: '#9196a5',
              fontWeight: '700',
            }}>
              {item.source.name.toUpperCase().slice(-4) === '.COM'
                ? item.source.name.toUpperCase().slice(0, -4).replace(/ /g, '')
                : item.source.name.toUpperCase().replace(/ /g, '')}
            </Text>
          </View>
          <View style={styles.timeStyle}>
            <Text style={{
              fontSize: 15,
              color: '#9196a5',
              fontWeight: '600',
            }}>{this.timePublished(item.publishedAt)}</Text>
          </View>
        </View>
      </TouchableOpacity>
    )
  }

  render() {
    if (this.state.news.length === 0) {
      return (
        <View style={styles.loadingStyle}>
          <MyBar
            height={Math.round(Dimensions.get('window').height)}
            width={Math.round(Dimensions.get('window').width)}
          />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <FlatList
          style={styles.flatStyle}
          data={this.state.news}
          renderItem={this.renderRow}
          keyExtractor={item => item.listId}
          refreshControl={
            <RefreshControl
              refreshing={this.state.refreshing}
              onRefresh={this.onRefresh}
            />
          }
        />
        {this.renderFilterSheet()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  sourceStyle: {
    flex: 1,
    alignItems: 'flex-start',
    paddingTop: 5,
  },
  timeStyle: {
    flex: 1,
    alignItems: 'flex-end',
  },
  loadingStyle: {
    flex: 1,
    backgroundColor: background,
    flexDirection: 'column',
    justifyContent: 'center',
  },
  row: {
    backgroundColor: '#515360',
    elevation: 1,
    borderRadius: 5,
    flexDirection: 'column',
    width: Math.round(Dimensions.get('window').width) - 28, 
    alignItems: 'center',
    paddingTop: 15,
    paddingBottom: 13,
    paddingHorizontal: 18,
    marginHorizontal: 14,
    marginTop: 0,
    marginBottom: 12,
  },
  imageContainer: {
    justifyContent: 'center',
  },
  imageStyle: {
    width: 80,
    height: 80,
  },
  textStyle: {
    flex: 1,
    paddingLeft: 18,
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: background,
  },
  filterContainer: {
    marginVertical: 12,
    marginHorizontal: 4,
    paddingBottom: 20,
  },
  cashText: {
    fontSize: 30,
    fontWeight: 'bold'
  },
  flatStyle: {
    marginTop: 20,
  },
  filterSheetHeader: {
    fontSize: 20,
    textAlign: "center",
    paddingBottom: 7,
    color: '#dbdbdb',
    fontWeight: '500',
  },
});
