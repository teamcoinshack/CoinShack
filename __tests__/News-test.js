/**
 * @format
 */

import 'react-native';
import React from 'react';
import News from '../screens/News.js';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('react-native-google-signin', () => {});
test('renders correctly', () => {
  const tree = renderer.create(<News />).toJSON();
  expect(tree).toMatchSnapshot();
});
