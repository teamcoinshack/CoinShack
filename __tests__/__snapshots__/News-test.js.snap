// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#373b48",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 1334,
        "justifyContent": "center",
        "width": 750,
      }
    }
  >
    <View
      animated={true}
      animationConfig={
        Object {
          "bounciness": 0,
        }
      }
      animationType="spring"
      indeterminate={true}
      indeterminateAnimationDuration={1000}
      onLayout={[Function]}
      progress={0.3}
      style={
        Array [
          Object {
            "backgroundColor": undefined,
            "borderColor": "#80e7a5",
            "borderRadius": 4,
            "borderWidth": 0,
            "overflow": "hidden",
            "width": 250,
          },
          undefined,
        ]
      }
      useNativeDriver={false}
    >
      <View
        style={
          Object {
            "backgroundColor": "#80e7a5",
            "height": 6,
            "transform": Array [
              Object {
                "translateX": 0,
              },
              Object {
                "translateX": -87.5,
              },
              Object {
                "scaleX": 0.30007,
              },
            ],
          }
        }
      />
    </View>
  </View>
</View>
`;
